system/Encryption/Encryption.php

```diff
@@ -0,0 +1,298 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @package    CodeIgniter
+ * @author     CodeIgniter Dev Team
+ * @copyright  2014-2017 British Columbia Institute of Technology (https://bcit.ca/)
+ * @license    https://opensource.org/licenses/MIT	MIT License
+ * @link       https://codeigniter.com
+ * @since      Version 4.0.0
+ * @filesource
+ */
+
+namespace CodeIgniter\Encryption;
+
+use Config\Encryption as EncryptionConfig;
+use CodeIgniter\Encryption\Exceptions\EncryptionException;
+use CodeIgniter\Config\Services;
+
+/**
+ * CodeIgniter Encryption Manager
+ *
+ * Provides two-way keyed encryption via PHP's Sodium and/or OpenSSL extensions.
+ * This class determines the driver, cipher, and mode to use, and then
+ * initializes the appropriate encryption handler.
+ */
+class Encryption
+{
+
+	/**
+	 * The encrypter we create
+	 *
+	 * @var string
+	 */
+	protected $encrypter;
+
+	/**
+	 * Our remembered configuration
+	 */
+	protected $config = null;
+
+	/**
+	 * Our default configuration
+	 */
+	protected $default = [
+		'driver' => 'OpenSSL', // The PHP extension we plan to use
+		'key'    => '', // no starting key material
+	];
+	protected $driver, $key;
+
+	/**
+	 * HMAC digest to use
+	 */
+	protected $digest = 'SHA512';
+
+	/**
+	 * Map of drivers to handler classes, in preference order
+	 *
+	 * @var array
+	 */
+	protected $drivers = [
+		'OpenSSL',
+	];
+
+	// --------------------------------------------------------------------
+
+	/**
+	 * Class constructor
+	 *
+	 * @param  mixed $params Configuration parameters
+	 * @return void
+	 *
+	 * @throws \CodeIgniter\Encryption\Exceptions\EncryptionException
+	 */
+	public function __construct($params = null)
+	{
+		$this->config = array_merge($this->default, (array) new \Config\Encryption());
+
+		if (is_string($params))
+		{
+			$params = ['driver' => $params];
+		}
+
+		$params = $this->properParams($params);
```

Comment:
I'm a bit confused why we're keeping an instance of `$params` here since they are set to `$this->config` in `properParams()`? 

Seems cleaner to do `$this->config = $this->properparams($params);` here and not have properParams set it to me.

Reply:
reworked